internal/util/basesubjectset.go

```diff
@@ -3,210 +3,230 @@ package util
+		bss.wildcard.unionWildcard(foundSubject)
+		bss.concrete.unionWildcard(foundSubject)
+		return !hasExisting
 	}
 
-	if bss.combiner != nil && ok {
-		bss.values[foundSubject.GetSubjectId()] = bss.combiner(bss.values[foundSubject.GetSubjectId()], foundSubject)
+	if len(foundSubject.GetExcludedSubjects()) > 0 {
+		panic("found excluded subjects for non-wildcard")
 	}
 
-	return !ok
+	_, hasExisting := bss.concrete.get(foundSubject.GetSubjectId())
+	bss.wildcard.unionConcrete(foundSubject)
+	bss.concrete.unionConcrete(foundSubject)
+	return !hasExisting
 }
 
 // Subtract subtracts the given subject found the set.
 func (bss BaseSubjectSet[T]) Subtract(foundSubject T) {
-	// If the subject being removed is a wildcard, then remove any non-excluded items and adjust
-	// the existing wildcard.
 	if foundSubject.GetSubjectId() == tuple.PublicWildcard {
-		exclusions := NewSet[string](foundSubject.GetExcludedSubjectIds()...)
-		for existingSubjectID := range bss.values {
-			if existingSubjectID == tuple.PublicWildcard {
-				continue
-			}
-
-			if !exclusions.Has(existingSubjectID) {
-				delete(bss.values, existingSubjectID)
-			}
-		}
+		bss.concrete.subtractWildcard(foundSubject)
 
-		// Check for an existing wildcard and adjust accordingly.
-		if existing, ok := bss.values[tuple.PublicWildcard]; ok {
-			// A subtraction of a wildcard from another wildcard subtracts the exclusions from the second.
-			// from the first, and places them into the subject set directly.
-			//
-			// As a concrete example, given `user:* - user:tom` - `user:* - user:sarah`, the subtraction
-			// of the two will be `user:sarah`, since sarah is in the first set and not in the second.
-			existingExclusions := NewSet[string](existing.GetExcludedSubjectIds()...)
-			for _, subjectID := range foundSubject.GetExcludedSubjectIds() {
-				if !existingExclusions.Has(subjectID) {
-					bss.values[subjectID] = bss.constructor(subjectID, nil, foundSubject)
-				}
-			}
+		// NOTE: subtracting a wildcard from another wildcard can result in concrete subjects
+		// being produced, which are added here, if any.
+		for _, subject := range bss.wildcard.subtractWildcard(foundSubject) {
+			bss.concrete.unionConcrete(subject)
 		}
-
-		delete(bss.values, tuple.PublicWildcard)
 		return
 	}
 
-	// Remove the subject itself from the set.
-	delete(bss.values, foundSubject.GetSubjectId())
-
-	// If wildcard exists within the subject set, add the found subject to the exclusion list.
-	if wildcard, ok := bss.values[tuple.PublicWildcard]; ok {
-		exclusions := NewSet[string](wildcard.GetExcludedSubjectIds()...)
-		exclusions.Add(foundSubject.GetSubjectId())
-		bss.values[tuple.PublicWildcard] = bss.constructor(tuple.PublicWildcard, exclusions.AsSlice(), wildcard)
+	if len(foundSubject.GetExcludedSubjects()) > 0 {
+		panic("found excluded subjects for non-wildcard")
 	}
+
+	bss.wildcard.subtractConcrete(foundSubject)
+	bss.concrete.subtractConcrete(foundSubject)
```

Comment:
I think this double-bookkeeping is weird but I don't have an alternate suggestion at the moment.

Every operation actually modifies these two underlying sets, and I'll have to figure out how they get combined elsewhere.

Reply:
The combination is just taking the values in each; they each track their own state independently.